src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -47,405 +49,429 @@
-  grpc_closure tcp_server_shutdown_complete;
-  grpc_closure* server_destroy_listener_done;
-  grpc_core::HandshakeManager* pending_handshake_mgrs;
-  grpc_core::RefCountedPtr<grpc_core::channelz::ListenSocketNode>
-      channelz_listen_socket;
-} server_state;
-
-typedef struct {
-  gpr_refcount refs;
-  server_state* svr_state;
-  grpc_pollset* accepting_pollset;
-  grpc_tcp_server_acceptor* acceptor;
-  grpc_core::RefCountedPtr<grpc_core::HandshakeManager> handshake_mgr;
-  // State for enforcing handshake timeout on receiving HTTP/2 settings.
-  grpc_chttp2_transport* transport;
-  grpc_millis deadline;
-  grpc_timer timer;
-  grpc_closure on_timeout;
-  grpc_closure on_receive_settings;
-  grpc_pollset_set* interested_parties;
-} server_connection_state;
-
-static void server_connection_state_unref(
-    server_connection_state* connection_state) {
-  if (gpr_unref(&connection_state->refs)) {
-    if (connection_state->transport != nullptr) {
-      GRPC_CHTTP2_UNREF_TRANSPORT(connection_state->transport,
-                                  "receive settings timeout");
-    }
-    grpc_pollset_set_del_pollset(connection_state->interested_parties,
-                                 connection_state->accepting_pollset);
-    grpc_pollset_set_destroy(connection_state->interested_parties);
-    gpr_free(connection_state);
+namespace grpc_core {
+namespace {
+
+class Chttp2ServerListener : public ServerListenerInterface {
+ public:
+  static grpc_error* Create(grpc_server* server, const char* addr,
+                            grpc_channel_args* args, int* port_num);
+
+  static grpc_error* CreateWithAcceptor(grpc_server* server, const char* name,
+                                        grpc_channel_args* args);
+
+  // Do not instantiate directly.  Use one of the factory methods above.
+  Chttp2ServerListener(grpc_server* server, grpc_channel_args* args);
+  ~Chttp2ServerListener();
+
+  void SetOnDestroyDone(grpc_closure* on_destroy_done) override;
+
+  void Start(grpc_server* server, grpc_pollset** pollsets, size_t npollsets)
+      override;
+
+  channelz::ListenSocketNode* channelz_listen_socket_node() const override {
+    return channelz_listen_socket_.get();
+  }
+
+  void Orphan() override;
+
+ private:
+  class ConnectionState : public RefCounted<ConnectionState> {
+   public:
+    ConnectionState(Chttp2ServerListener* listener,
+                    grpc_pollset* accepting_pollset,
+                    grpc_tcp_server_acceptor* acceptor,
+                    RefCountedPtr<HandshakeManager> handshake_mgr,
+                    grpc_channel_args* args, grpc_endpoint* endpoint);
+
+    ~ConnectionState();
+
+   private:
+    static void OnTimeout(void* arg, grpc_error* error);
+    static void OnReceiveSettings(void* arg, grpc_error* error);
+    static void OnHandshakeDone(void* arg, grpc_error* error);
+
+    Chttp2ServerListener* listener_;
+    grpc_pollset* accepting_pollset_;
+    grpc_tcp_server_acceptor* acceptor_;
```

Comment:
```suggestion
    Chttp2ServerListener* const listener_;
    grpc_pollset* const accepting_pollset_;
    grpc_tcp_server_acceptor* const acceptor_;
```

to emphasize a lack of change in these and that they don't need a default initializer